### PR TITLE
Terraform: Access PostgreSQL server through private link

### LINKED ISSUE #1728: Terraform: Connect application to database through Private Link
See: https://docs.microsoft.com/en-us/azure/postgresql/concepts-data-access-and-security-private-link

### PR DESCRIPTION
**Pull request by shubha-rajan**
### Description
Change database configuration to block public access to Postgres server and set up private link.

Note: This will need to be rebased and merged after #1726 is merged

Fixes #1728

### CODE REVIEW

**Review by zaralouis-sf — APPROVED** [review]

File: cloud/azure/main.tf [hunk 1]
```diff
@@ -168,7 +167,7 @@ resource "azurerm_postgresql_server" "civiform" {
   auto_grow_enabled            = true
 
   # TODO: configure a subnet and restrict access only to the application servers.
-  public_network_access_enabled = true
+  public_network_access_enabled = false
```
Comment (bion) [inline cloud/azure/main.tf:170]: question (blocking): does this PR resolve this TODO?

File: cloud/azure/main.tf [hunk 2]
```diff
@@ -182,10 +181,43 @@ resource "azurerm_postgresql_database" "civiform" {
...
+  enforce_private_link_endpoint_network_policies = true
+}
+
+resource "azurerm_private_dns_zone" "privatelink" {
+  name                = "privatelink.postgres.database.azure.com"
+  resource_group_name = azurerm_resource_group.rg.name
+}
+
+resource "azurerm_private_dns_zone_virtual_network_link" "vnet_link" {
+  name                  = "vnet_link"
```
Comment (bion) [inline cloud/azure/main.tf:200]: suggestion (non-blocking): would be good to have a consistent naming convention for the Azure-side names of things. Maybe changing this one to match the convention used on L195?

File: cloud/azure/outputs.tf [hunk 3]
```diff
@@ -1,3 +1,7 @@
 output "app_service_default_hostname" {
   value = "https://${azurerm_app_service.civiform_app.default_site_hostname}"
 }
+
+output "pg_server_privatelink_fqdn" {
+  value = azurerm_private_endpoint.endpoint.private_dns_zone_configs[0].record_sets[0].fqdn
```
Comment (bion) [inline cloud/azure/outputs.tf:6]: question (blocking): does terraform support storing `azurerm_private_endpoint.endpoint.private_dns_zone_configs[0].record_sets[0].fqdn` in a variable so the variable can be referenced here and on L75 of main.tf?

File: cloud/azure/main.tf (hunk 1 above)
Comment (shubha-rajan) [reply cloud/azure/main.tf]: Pretty sure it does! It restricts access to only other servers inside the VNET.  Removed the comment.

File: cloud/azure/main.tf (hunk 2 above)
Comment (shubha-rajan) [reply cloud/azure/main.tf]: We can try to standardize these as much as possible, but each resource has its own naming rules. Database servers and App Service webapps for example only accept alphanumeric characters and hyphens. Storage accounts only accept alphanumeric characters. On L195, the name had to be a valid domain name: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/private_dns_zone

File: cloud/azure/outputs.tf (hunk 3 above)
Comment (shubha-rajan) [reply cloud/azure/outputs.tf]: It does. That's what [local values](https://www.terraform.io/docs/language/values/locals.html) are used for. Added this to `locals.tf`

File: cloud/azure/main.tf [hunk 4]
```diff
@@ -73,7 +72,7 @@ resource "azurerm_app_service" "civiform_app" {
 
     DB_USERNAME    = "${azurerm_postgresql_server.civiform.administrator_login}@${azurerm_postgresql_server.civiform.name}"
     DB_PASSWORD    = azurerm_postgresql_server.civiform.administrator_login_password
-    DB_JDBC_STRING = "jdbc:postgresql://${azurerm_postgresql_server.civiform.fqdn}:5432/postgres?ssl=true&sslmode=require"
+    DB_JDBC_STRING = "jdbc:postgresql://${azurerm_private_endpoint.endpoint.private_dns_zone_configs[0].record_sets[0].fqdn}:5432/postgres?ssl=true&sslmode=require"
```
[the remark this answers is not in the document]
Comment (shubha-rajan) [reply cloud/azure/main.tf]: done

File: cloud/azure/main.tf (hunk 1 above)
Comment (bion) [reply cloud/azure/main.tf]: nice

File: cloud/azure/main.tf (hunk 2 above)
Comment (bion) [reply cloud/azure/main.tf]: 😭

thanks for looking into this

**Review by bion — APPROVED** [review]

**Pull request merged**